clang/lib/Sema/SemaExpr.cpp

```diff
@@ -229,6 +229,20 @@ bool Sema::DiagnoseUseOfDecl(NamedDecl *D, ArrayRef<SourceLocation> Locs,
                !checkAllowedSYCLInitializer(VD, /*CheckValueDependent =*/true))
         SYCLDiagIfDeviceCode(*Locs.begin(), diag::err_sycl_restrict)
             << Sema::KernelConstStaticVariable;
+      else if (IsConst && VD->hasGlobalStorage()) {
+        QualType Ty = VD->getType();
+        if (const auto *CAType = dyn_cast<ConstantArrayType>(Ty)) {
+          assert(CAType);
+          // Max length of SPIRV instruction is 65535 words.
+          // First 3 words of OpConstantComposite encode: 1) word count &
+          // opcode, 2) Result Type and 3) Result Id.
+          const int MaxNumElements = 65535 - 3;
```

Comment:
Why can't this just be 65532?

Reply:
Having it as `65535 - 3;` vs just `65532` (along with the comment above it) is to give some visual context of the logic explained below.
OpConstantComposite instruction takes a variable number of operands. 
For example, if we have an array with 65535 elements , OpConstantComposite  instruction will have 65535 operands. 
Each operand takes one word. 
Per SPIR-V spec, WordCount limit for this instruction is 3 + number of operand it takes.
So the word count for OpConstantComposite will be 3 + 65535 = 65538
Also, Word Count is the high-order 16 bits of word 0 of the instruction, holding its total WordCount. 
65538 is more then the maximum value which can be stored in the high-order 16-bits. 
Hence the subtraction of 3 words from the minimum allowed limit.

Reply:
Right. But I don't think this is relevant from a FE perspective. I think a simple comment along the lines of array size cannot exceed 65532 due to SPIR-V limitations suffices.

Reply:
Agree, I can update the comment to be FE friendly.